docker/service.go

```diff
@@ -344,21 +358,58 @@ func (s *Service) collectContainersAndDo(ctx context.Context, action func(*Conta
 }
 
 func (s *Service) eachContainer(ctx context.Context, containers []*Container, action func(*Container) error) error {
-
 	tasks := utils.InParallel{}
 	for _, container := range containers {
-		task := func(container *Container) func() error {
-			return func() error {
-				return action(container)
-			}
-		}(container)
+		task := taskFunc(container, action)
+		tasks.Add(task)
+	}
+	return tasks.Wait()
+}
+
+func (s *Service) getFilters() filters.Args {
+	filter := filters.NewArgs()
+	filter.Add("label", fmt.Sprintf("%s=%s", labels.PROJECT, s.project.Name))
+	filter.Add("label", fmt.Sprintf("%s=%s", labels.SERVICE, s.name))
+	return filter
+}
 
+func (s *Service) watchContainers(ctx context.Context, initialContainers []*Container, action func(*Container) error, dieAction func(*Container) error) error {
```

Comment:
May be some more events may require to be watched like ``start`` etc.
I don't have use case of other event then ``start`` right now, but may be in future it may require. Will it be better to pass list of ``event , handler`` pair instead of predefined event-handlers.


One example:
While implementing ``--abort-on-container-exit`` in ``Up`` command, I need to track if all containers have started or died. To keep track I need to watch ``start`` event.